nixos/modules/services/misc/nzbget.nix

```diff
@@ -64,10 +96,27 @@ in
         unrar
         p7zip
       ];
-      preStart = ''
+      preStart = if cfg.declarativeConfig == null then ''
         if [ ! -f ${configFile} ]; then
           ${pkgs.coreutils}/bin/install -m 0700 ${pkg}/share/nzbget/nzbget.conf ${configFile}
         fi
+      '' else ''
+        set -u
```

Comment:
This is a bit scary...

Reply:
I understand. I tried simply concatenating the config files first, but this did somehow not work and earlier options seemed to take precedence. To remove duplicate options this should behave like NZBGet's config parser (see <https://github.com/nzbget/nzbget/blob/develop/daemon/main/Options.cpp#L1414>).

Reply:
If you have an existing `nzbget` install you already have secrets stored in the file and you're good to go. If you want to deploy a new instance you can place a **minimal** copy of your `nzbget.conf` file under `/var/lib/nzbget` and declaratively configure the rest with new `settings` option in #132722. This would work nicely and keeps the code **very** simple.

Reply:
I am currently trying to fully declare my nzbget config and store my secrets with SOPs. The soution @aanderse proposed above setting `/var/lib/nzbget/nzbget.conf` would work great for me. I can store my few sensitive values in a text file, and declare the rest via the `settings` option.

However, it appears [the `config` option was removed in `5b76046db38124ef8cf06dad4dce9e0765ad8ff5`.](https://github.com/NixOS/nixpkgs/commit/5b76046db38124ef8cf06dad4dce9e0765ad8ff5#diff-c3dc41382e07d8dca09bfd95714681909e53d277d6226dec72c0dfbf2d04d399).

Would it be reasonable to reintroduce that option? I'd be happy to put in the PR for it if so.

Would it be a reasonable idea to expose a `configFile`

https://github.com/NixOS/nixpkgs/commit/4ee3e8b21dfeaf236b86a0c109953dd809a77e64

Reply:
@KyleOndy just place your configuration file containing all your secrets under `/var/lib/nzbget`, call it `nzbget.conf` and you're good to go. What do you need the `configFile` option for?

Reply:
> place your configuration file containing all your secrets under /var/lib/nzbget

Agreed that would work. Is it possible to provision a file outside of `/etc` with NixOS? That is the crux of the issue. I'd like to have my config fully declared without having to add in logic to build the config file with secrets and move it to the host. 

Using [sops-nix](https://github.com/Mic92/sops-nix) I'm able to pass my secret config via a file, `configFile = config.sops.secrets."nzbget/config".path;`, and configure everything else with the `settings` option. 

---
Here is the actual implementation at this point in time. These may break in the future since this is a branch under heavy work.

The import diff of `nzbget` is:

```diff
diff --git 1/home/kyle/src/nixpkgs/nixos/modules/services/misc/nzbget.nix 2/./modules/system_modules/nzbget.nix
index 27c5f2e..b90df45 100644
--- 1/home/kyle/src/nixpkgs/nixos/modules/services/misc/nzbget.nix
+++ 2/./modules/system_modules/nzbget.nix
@@ -3,10 +3,9 @@
 with lib;
 
 let
-  cfg = config.services.nzbget;
+  cfg = config.systemFoundry.services.nzbget;
   pkg = pkgs.nzbget;
   stateDir = "/var/lib/nzbget";
-  configFile = "${stateDir}/nzbget.conf";
   configOpts = concatStringsSep " " (mapAttrsToList (name: value: "-o ${name}=${escapeShellArg (toStr value)}") cfg.settings);
   toStr = v:
     if v == true then "yes"
@@ -15,16 +14,8 @@ let
     else v;
 in
 {
-  imports = [
-    (mkRemovedOptionModule [ "services" "misc" "nzbget" "configFile" ] "The configuration of nzbget is now managed by users through the web interface.")
-    (mkRemovedOptionModule [ "services" "misc" "nzbget" "dataDir" ] "The data directory for nzbget is now /var/lib/nzbget.")
-    (mkRemovedOptionModule [ "services" "misc" "nzbget" "openFirewall" ] "The port used by nzbget is managed through the web interface so you should adjust your firewall rules accordingly.")
-  ];
-
-  # interface
-
   options = {
-    services.nzbget = {
+    systemFoundry.services.nzbget = {
       enable = mkEnableOption "NZBGet";
 
       user = mkOption {
@@ -39,13 +30,19 @@ in
         description = "Group under which NZBGet runs";
       };
 
+      configFile = mkOption {
+        type = types.path;
+        default = "${stateDir}/nzbget.conf";
+        description = "Path to config file. Useful for reading secrets from disk";
+      };
+
       settings = mkOption {
         type = with types; attrsOf (oneOf [ bool int str ]);
-        default = {};
+        default = { };
         description = ''
           NZBGet configuration, passed via command line using switch -o. Refer to
           <link xlink:href="https://github.com/nzbget/nzbget/blob/master/nzbget.conf"/>
-          for details on supported values.
+          for details on supported values. These values override the `configFile` option.
         '';
         example = {
           MainDir = "/data";
@@ -57,7 +54,7 @@ in
   # implementation
 
   config = mkIf cfg.enable {
-    services.nzbget.settings = {
+    systemFoundry.services.nzbget.settings = {
       # allows nzbget to run as a "simple" service
       OutputMode = "loggable";
       # use journald for logging
@@ -83,8 +80,8 @@ in
       ];
 
       preStart = ''
-        if [ ! -f ${configFile} ]; then
-          ${pkgs.coreutils}/bin/install -m 0700 ${pkg}/share/nzbget/nzbget.conf ${configFile}
+        if [ ! -f ${cfg.configFile} ]; then
+          ${pkgs.coreutils}/bin/install -m 0700 ${pkg}/share/nzbget/nzbget.conf ${cfg.configFile}
         fi
       '';
 
@@ -95,7 +92,7 @@ in
         Group = cfg.group;
         UMask = "0002";
         Restart = "on-failure";
-        ExecStart = "${pkg}/bin/nzbget --server --configfile ${stateDir}/nzbget.conf ${configOpts}";
+        ExecStart = "${pkg}/bin/nzbget --server --configfile ${cfg.configFile} ${configOpts}";
         ExecStop = "${pkg}/bin/nzbget --quit";
       };
     };
```

- [Usage of the `configFile` option](https://github.com/KyleOndy/dotfiles/blob/9adf77d48150de9c926808545c617ec6bb78ebfd/hosts/util_dmz/configuration.nix#L36-L62)
- [version of `nzbget.nix` with the option reintroduced](https://github.com/KyleOndy/dotfiles/blob/home-infra/modules/system_modules/nzbget.nix)

Reply:
💡 That should would work. I'm still getting used to some of the patterns in NixOS. So I should be able to do something like the following:

```nix
services.nzbget = {
  config = {
    # non sensitive things
    Foo = "Bar";
    Bizz = "Baz";
  };
};

# since the above instantiates the systemd service, we're able to configure and override that service.
systemd.services.nzbget.preStart = lib.mkForce ''
  if [ ! -f /var/lib/nzbget/nzbget.conf ]; then
    ${pkgs.coreutils}/bin/install -m 0700 ${config.sops.secrets."nzbget/config".path} /var/lib/nzbget/nzbget.conf
  fi
'';
```

Reply:
Maybe your confusion stems from this software treating configuration as state. NixOS doesn't like this and doesn't work well with it. The configuration file is state because the web app is meant to mutate it. This module is basically a productive and practical hack, not pure nix philosophy :smile:

It is a "bad idea" to do what you want from the pure nix philosophy perspective because you want to manage state in a declarative way... but it is a fantastic idea from a pragmatic point of view, so go for it!

Does that make sense at all?

Reply:
No problem at all. Please never hesitate to ask in this community! I'm glad I could help.